src/tracer/src/cli/handlers/init/arguments.rs

```diff
@@ -48,28 +52,77 @@ pub struct TracerCliInitArgs {
     pub dev: bool,
 }
 
+#[derive(Debug, Default, Clone, PartialEq, ValueEnum)]
 pub enum PromptMode {
-    Always,
-    Never,
-    WhenMissing,
+    /// do not prompt (i.e. non-interactive)
+    None,
+    /// only prompt for minimal information - automatically generate missing values when possible
+    #[default]
+    Minimal,
+    /// prompt for all required values
+    Required,
+    // /// prompt for all values
+    // All,
 }
 
 impl TracerCliInitArgs {
-    pub fn finalize(self, default_prompt_mode: PromptMode) -> FinalizedInitArgs {
-        let prompt_mode = if self.non_interactive {
-            PromptMode::Never
+    /// Fill in any missing arguments according to the `PromptMode`.
+    /// If `confirm` is `true`, then the user will be prompted for all options, and any specified
+    /// values will be used as defaults; otherwise, the user is only be for missing values.
+    pub async fn finalize(self, default_pipeline_prefix: &str, confirm: bool) -> FinalizedInitArgs {
+        let prompt_mode = if self.no_daemonize {
+            PromptMode::None
         } else {
-            default_prompt_mode
+            self.interactive_prompts
         };
+        let mut tags = self.tags;
+
+        // user_id is required - try to get it from the command line, fall back to user prompt
+        // unless mode is set to non-interactive; error if missing
+        let user_id = match (tags.user_id, &prompt_mode) {
+            (Some(user_id), PromptMode::Minimal | PromptMode::Required) if confirm => {
+                Some(Self::prompt_for_user_id(Some(&user_id)))
+            }
+            (Some(user_id), _) => Some(user_id),
+            (None, PromptMode::Minimal | PromptMode::Required) => {
+                Some(Self::prompt_for_user_id(None))
+            }
+            (None, PromptMode::None) => None,
+        }
+        // TODO: remove this once we can source the user ID from the credentials file
+        .or_else(|| {
```

Comment:
I would add the environment variable reading after the command line input, if there is no cli parameter `--user-id` or there is no environment variable detected, then we show the prompt

Reply:
You're right, I had thought using `env` with `clap` would read the environment variable only when the command line option does not exist, but it preferences the environment variable over the command line option.

I will switch back to manually reading environment variables if the command line option does not exist.